engine/commands/cortex_upd_cmd.cc

```diff
@@ -12,7 +12,182 @@
+          return std::nullopt;
+        }
+        std::string current_version = CORTEX_CPP_VERSION;
+        CTL_INF("Got the latest release, update to the config file: "
+                << latest_version)
+        config.latestRelease = latest_version;
+        config_yaml_utils::DumpYamlConfig(
+            config, file_manager_utils::GetConfigurationPath().string());
+        if (current_version != latest_version) {
+          return latest_version;
+        }
+      } catch (const nlohmann::json::parse_error& e) {
+        CTL_INF("JSON parse error: " << e.what());
+        return std::nullopt;
+      }
+    } else {
+      CTL_INF("HTTP error: " << res->status);
+      return std::nullopt;
+    }
+  } else {
+    auto err = res.error();
+    CTL_INF("HTTP error: " << httplib::to_string(err));
+    return std::nullopt;
+  }
+  return std::nullopt;
+}
+
+bool ReplaceBinaryInflight(const std::filesystem::path& src,
+                           const std::filesystem::path& dst) {
+  if (src == dst) {
+    // Already has the newest
+    return true;
+  }
+
+  std::filesystem::path temp = dst.parent_path() / "cortex_temp";
+  auto restore_binary = [&temp, &dst]() {
+    if (std::filesystem::exists(temp)) {
+      std::rename(temp.string().c_str(), dst.string().c_str());
+      CLI_LOG("Restored binary file");
+    }
+  };
+
+  try {
+    if (std::filesystem::exists(temp)) {
+      std::filesystem::remove(temp);
+    }
+#if !defined(_WIN32)
+    // Get permissions of the executable file
+    struct stat dst_file_stat;
+    if (stat(dst.string().c_str(), &dst_file_stat) != 0) {
+      CTL_ERR("Error getting permissions of executable file: " << dst.string());
+      return false;
+    }
+
+    // Get owner and group of the executable file
+    uid_t dst_file_owner = dst_file_stat.st_uid;
+    gid_t dst_file_group = dst_file_stat.st_gid;
+#endif
+
+    std::rename(dst.string().c_str(), temp.string().c_str());
+    std::filesystem::copy_file(
+        src, dst, std::filesystem::copy_options::overwrite_existing);
+
+#if !defined(_WIN32)
+    // Set permissions of the executable file
+    if (chmod(dst.string().c_str(), dst_file_stat.st_mode) != 0) {
+      CTL_ERR("Error setting permissions of executable file: " << dst.string());
+      restore_binary();
+      return false;
+    }
+
+    // Set owner and group of the executable file
+    if (chown(dst.string().c_str(), dst_file_owner, dst_file_group) != 0) {
+      CTL_ERR(
+          "Error setting owner and group of executable file: " << dst.string());
+      restore_binary();
+      return false;
+    }
+
+    // Remove cortex_temp
+    if (unlink(temp.string().c_str()) != 0) {
+      CTL_ERR("Error deleting self: " << strerror(errno));
+      restore_binary();
+      return false;
+    }
+#endif
+  } catch (const std::exception& e) {
+    CTL_ERR("Something went wrong: " << e.what());
+    restore_binary();
+    return false;
+  }
+
+  return true;
+}
+
 void CortexUpdCmd::Exec(std::string v) {
```

Comment:
just a question, why not `const std::string& v`?

Reply:
Nice catch!  Seems like we intended to use std::move() in early day. Will fix it